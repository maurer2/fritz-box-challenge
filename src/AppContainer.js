import React, { Component } from 'react';

import { getTimeBetween, getDate, getDateAsIsoDate, getNowDate } from './libs/time';
import { getMappedFields } from './libs/mapper';
import { transformString as splitData, getDashPositonsInString, splitString as splitToArray } from './libs/splitter';
import getData from './libs/modem';
import parseData from './libs/parser';

import mockResponse from './mocks/box-01004.txt';

import App from './App';

class AppContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isUpdating: true,
      isValid: true,
      url: (process.env.REACT_APP_MODE === 'dev') ? mockResponse : '/cgi-bin/system_status',
      boxData: {},
      componentsToShow: ['branding', 'firmware', 'model', 'restarts', 'technology', 'runtime', 'age'],
      currentIndex: 0,
    };

    this.handleClick = this.handleClick.bind(this);
    this.handleNavigation = this.handleNavigation.bind(this);
  }

  handleClick() {
    this.setState((previousState) => {
      const { currentIndex, componentsToShow } = previousState;
      const lastIndex = componentsToShow.length - 1;

      return {
        currentIndex: (currentIndex < lastIndex) ? currentIndex + 1 : 0,
      };
    });
  }

  handleNavigation(index) {
    this.setState({ currentIndex: index });
  }

  mapBoxData(boxData, runtime, age) {
    const { componentsToShow } = this.state;

    const mappedEntries = componentsToShow.reduce((total, current) => {
      const entries = total;

      entries[current] = boxData[current] || '';

      return entries;
    }, {});

    mappedEntries.runtime = runtime;
    mappedEntries.age = age;

    return mappedEntries;
  }

  getBoxData() {
    this.setState({ isUpdating: true });

    const fetchedFinally = getData(this.state.url)
      .then((data) => {
        const parsedTextString = parseData(data);

        const dashPositions = getDashPositonsInString(parsedTextString);
        const splitString = splitData(parsedTextString, dashPositions);
        const splitStringAsArray = splitToArray(splitString);

        const mappedValues = getMappedFields(splitStringAsArray);

        const extractedDateString = `${mappedValues['powerOnHours 1']}-${mappedValues['powerOnHours 2']}`;
        const nowDateString = getNowDate();

        const dateIsoString = getDateAsIsoDate(extractedDateString, nowDateString);
        const runtime = getDate(dateIsoString);
        const age = getTimeBetween(dateIsoString, nowDateString);

        this.setState(() => {
          const boxData = this.mapBoxData(mappedValues, runtime, age);

          return {
            boxData,
            isUpdating: false,
            isValid: true,
          };
        });
      })
      .catch(() => {
        this.setState({ isValid: false });

        Promise.resolve();
      });

    fetchedFinally.then(() => {
      this.setState({ isUpdating: false });
    });
  }

  componentDidMount() {
    this.getBoxData();
  }

  render() {
    const { isUpdating, boxData, currentIndex, isValid } = this.state;

    return (
      <App
        isUpdating={ isUpdating }
        isValid={ isValid }
        boxData={ boxData }
        currentIndex={ currentIndex }
        handleClickEvent={ this.handleClick }
        handleNavigation={ this.handleNavigation }
      />
    );
  }
}

export default AppContainer;
