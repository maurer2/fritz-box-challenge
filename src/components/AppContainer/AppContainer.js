import React, { useState, useEffect } from 'react';

import { getTimeBetween, getDate, getDateAsIsoDate, getNowDate } from '../../libs/time';
import { getMappedFields } from '../../libs/mapper';
import { transformString as splitData, getDashPositonsInString, splitString as splitToArray } from '../../libs/splitter';
import getData from '../../libs/modem';
import parseData from '../../libs/parser';

import mockResponse from '../../mocks/box-iu7nl.txt';

import { App } from '../App';

function mapBoxData(componentsToShow, boxData, runtime, age) {
  const mappedEntries = componentsToShow.reduce((total, current) => {
    const entries = total;

    entries[current] = boxData[current] || '';

    return entries;
  }, {});

  mappedEntries.runtime = runtime;
  mappedEntries.age = age;

  return mappedEntries;
}

const AppContainer = () => {
  const [isUpdating, setIsUpdating] = useState(true);
  const [isValid, setIsValid] = useState(true);
  const [boxData, setBoxData] = useState({});

  const url = (process.env.REACT_APP_MODE === 'dev') ? mockResponse : '/cgi-bin/system_status';
  const componentsToShow = ['branding', 'firmware', 'model', 'restarts', 'technology', 'runtime', 'age'];

  useEffect(() => {
    // eslint-disable-next-line
    getBoxData();
  // eslint-disable-next-line
  }, []);

  function getBoxData() {
    setIsUpdating(true);

    const fetchedFinally = getData(url)
      .then((data) => {
        const parsedTextString = parseData(data);

        const dashPositions = getDashPositonsInString(parsedTextString);
        const splitString = splitData(parsedTextString, dashPositions);
        const splitStringAsArray = splitToArray(splitString);

        const mappedValues = getMappedFields(splitStringAsArray);

        const extractedDateString = `${mappedValues['powerOnHours 1']}-${mappedValues['powerOnHours 2']}`;
        const nowDateString = getNowDate();

        const dateIsoString = getDateAsIsoDate(extractedDateString, nowDateString);
        const runtime = getDate(dateIsoString);
        const age = getTimeBetween(dateIsoString, nowDateString);

        const newBoxData = mapBoxData(componentsToShow, mappedValues, runtime, age);

        setBoxData(newBoxData);
        setIsValid(true);
      })
      .catch(() => {
        setIsValid(false);

        Promise.resolve();
      });

    fetchedFinally.then(() => {
      setIsUpdating(false);
    });
  }

  return (
    <App
      isUpdating={ isUpdating }
      isValid={ isValid }
      boxData={ boxData }
      componentsToShow={ componentsToShow}
    />
  );
};

export { AppContainer };
