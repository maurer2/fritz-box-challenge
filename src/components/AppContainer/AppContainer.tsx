import React, { useState, useEffect } from 'react';

import {
  getTimeBetween, getDate, getDateAsIsoDate, getNowDate,
} from '../../libs/time';
import { getMappedFields } from '../../libs/mapper';
import {
  transformString as splitData,
  getDashPositionsInString,
  splitString as splitToArray,
} from '../../libs/splitter';
import getData from '../../libs/modem';
import parseData from '../../libs/parser';

import mockResponse from '../../mocks/box-iu7nl.txt';

import * as Types from './AppContainer.types';

import { App } from '../App';

function mapBoxData(componentsToShow: any, boxData: any, runtime: any, age: any): Types.StringMap {
  const mappedEntries = componentsToShow.reduce((total: any, current: any) => {
    const entries = total;

    entries[current] = boxData[current] || '';

    return entries;
  }, {} as any);

  mappedEntries.runtime = runtime;
  mappedEntries.age = age;

  return mappedEntries;
}

const AppContainer: React.FC = (): JSX.Element => {
  const [isUpdating, setIsUpdating] = useState(true);
  const [isValid, setIsValid] = useState(true);
  const [boxData, setBoxData] = useState({});

  const url = process.env.REACT_APP_MODE === 'dev' ? mockResponse : '/cgi-bin/system_status';
  const componentsToShow = [
    'branding',
    'firmware',
    'model',
    'restarts',
    'technology',
    'runtime',
    'age',
  ];

  function getBoxData(): void {
    setIsUpdating(true);

    const fetchedFinally = getData(url)
      .then((data: any) => {
        const parsedTextString = parseData(data);

        const dashPositions = getDashPositionsInString(parsedTextString);
        const splitString = splitData(parsedTextString, dashPositions);
        const splitStringAsArray = splitToArray(splitString);

        const mappedValues = getMappedFields(splitStringAsArray);

        const extractedDateString = `${mappedValues['powerOnHours 1']}-${mappedValues['powerOnHours 2']}`;
        const nowDateString = getNowDate();

        const dateIsoString = getDateAsIsoDate(extractedDateString, nowDateString);

        const runtime = getDate(dateIsoString);

        const age = getTimeBetween(dateIsoString, nowDateString);

        const newBoxData = mapBoxData(componentsToShow, mappedValues, runtime, age);

        setBoxData(newBoxData);
        setIsValid(true);
      })
      .catch((error: Error) => {
        console.log(error);
        setIsValid(false);

        Promise.resolve();
      });

    fetchedFinally.then(() => {
      setIsUpdating(false);
    });
  }

  useEffect(() => {
    getBoxData();
    // eslint-disable-next-line
  }, []);

  return (
    <App
      isUpdating={isUpdating}
      isValid={isValid}
      boxData={boxData}
      componentsToShow={componentsToShow}
    />
  );
};

export { AppContainer };
