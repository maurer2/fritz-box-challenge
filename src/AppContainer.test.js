import React from 'react';
import ReactDOM from 'react-dom';
import AppContainer from './AppContainer';

it('renders without crashing', () => {
  const div = document.createElement('div');

  expect(true).toEqual(true);
  // ReactDOM.render(<AppContainer />, div);
  // ReactDOM.unmountComponentAtNode(div);
});
