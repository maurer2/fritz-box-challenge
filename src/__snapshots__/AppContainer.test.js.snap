// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer should match snapshot 1`] = `
<App
  boxData={Object {}}
  currentIndex={0}
  handleClickEvent={[Function]}
  handleNavigation={[Function]}
  isUpdating={false}
  isValid={false}
/>
`;
